HPCC-33678 Optimize Ptree deserialization - Step 1

streeterd opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

streeterd committed:
HPCC-33678 Optimize Ptree deserialization

Replace StringAttr with StringBuffer
Add unit test for deserialization timings

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd commented:
this is a draft PR as after making the changes to use StringBuffer I am seeing only very slight improvement in performance:
```
.16:32:03.813814 82231 UNK UNK Test Small Tree (wide): 100 nodes, depth 1, text content, 100000 iterations, StringAttr true     Tree size: 7502 bytes serialized        results: 17.23 s runtime
16:32:22.755654 82231 UNK UNK Test Small Tree (wide): 100 nodes, depth 1, text content, 100000 iterations, StringAttr false     Tree size: 7502 bytes serialized        results: 17.06 s runtime
16:32:40.936742 82231 UNK UNK Test Small Tree (deep): 100 nodes, depth 10, text content, 100000 iterations, StringAttr true     Tree size: 4812 bytes serialized        results: 16.34 s runtime
16:32:59.168333 82231 UNK UNK Test Small Tree (deep): 100 nodes, depth 10, text content, 100000 iterations, StringAttr false    Tree size: 4812 bytes serialized        results: 16.36 s runtime
.16:33:22.715621 82231 UNK UNK Test Medium Tree (wide): 1000 nodes, depth 1, text content, 10000 iterations, StringAttr true    Tree size: 75902 bytes serialized       results: 21.64 s runtime
16:33:45.699449 82231 UNK UNK Test Medium Tree (wide): 1000 nodes, depth 1, text content, 10000 iterations, StringAttr false    Tree size: 75902 bytes serialized       results: 21.08 s runtime
16:34:04.424544 82231 UNK UNK Test Medium Tree (deep): 1000 nodes, depth 20, text content, 10000 iterations, StringAttr true    Tree size: 48992 bytes serialized       results: 16.97 s runtime
16:34:23.183773 82231 UNK UNK Test Medium Tree (deep): 1000 nodes, depth 20, text content, 10000 iterations, StringAttr false   Tree size: 48992 bytes serialized       results: 17.01 s runtime
.16:35:31.611529 82231 UNK UNK Test Large Tree (wide): 10000 nodes, depth 1, text content, 1000 iterations, StringAttr true     Tree size: 768902 bytes serialized      results: 65.59 s runtime
16:36:37.485168 82231 UNK UNK Test Large Tree (wide): 10000 nodes, depth 1, text content, 1000 iterations, StringAttr false     Tree size: 768902 bytes serialized      results: 63.17 s runtime
16:36:44.490617 82231 UNK UNK Test Large Tree with binary (wide): 10000 nodes, depth 1, binary content, 100 iterations, StringAttr true Tree size: 1006902 bytes serialized     results: 6.60 s runtime
16:36:51.467253 82231 UNK UNK Test Large Tree with binary (wide): 10000 nodes, depth 1, binary content, 100 iterations, StringAttr false        Tree size: 1006902 bytes serializedresults: 6.55 s runtime
.16:37:00.710365 82231 UNK UNK Test Deep Tree: 5000 nodes, depth 500, text content, 1000 iterations, StringAttr true    Tree size: 245292 bytes serialized      results: 8.12 s runtime
16:37:10.134183 82231 UNK UNK Test Deep Tree: 5000 nodes, depth 500, text content, 1000 iterations, StringAttr false    Tree size: 245292 bytes serialized      results: 8.28 s runtime
16:37:32.624622 82231 UNK UNK Test Wide Tree: 5000 nodes, depth 1, text content, 1000 iterations, StringAttr true       Tree size: 383902 bytes serialized      results: 21.21 s runtime
16:37:55.050625 82231 UNK UNK Test Wide Tree: 5000 nodes, depth 1, text content, 1000 iterations, StringAttr false      Tree size: 383902 bytes serialized      results: 21.13 s runtime
16:38:05.227278 82231 UNK UNK Test Balanced Tree: 5000 nodes, depth 70, text content, 1000 iterations, StringAttr true  Tree size: 250622 bytes serialized      results: 9.28 s runtime
16:38:15.668746 82231 UNK UNK Test Balanced Tree: 5000 nodes, depth 70, text content, 1000 iterations, StringAttr false Tree size: 250622 bytes serialized      results: 9.52 s runtime
```
I looked at removing the StringBuffer clear() and creating a new read() function to use StringBuffer set() but set() uses clear() and append() which effectively is what my changes are using!

Any ideas for an improvement?

streeterd commented:
Comparing the deserialization using `StringBuffer `to replace `StringAttr` the following functions are used which show that the actual format does not change:

```
system/jlib/jptree.cpp

	void PTree::deserialize(MemoryBuffer &src)
	{
	    deserializeSelf(src);
	    StringAttr eName;
	    for (;;)
	    {
	        size32_t pos = src.getPos();
	        src.read(eName);
	        if (eName.isEmpty())
	            break;
	        src.reset(pos); // reset to re-read tree name
	        IPropertyTree *child = create(src);
	        addPropTree(eName, child);
	    }
	}
	
	
	void PTree::deserialize(MemoryBuffer &src)
	{
	    DeserializeContext deserializeContext;
	    deserializeSelf(src, deserializeContext);
	    for (;;)
	    {
	        size32_t pos = src.getPos();
	        deserializeContext.name.clear();
	        src.read(deserializeContext.name);
	        if (deserializeContext.name.isEmpty())
	            break;
	        src.reset(pos); // reset to re-read tree name
	        IPropertyTree *child = create(src);
	        addPropTree(deserializeContext.name, child);
	    }
	}


system/jlib/jbuff.cpp
	
	MemoryBuffer & MemoryBuffer::read(StringBuffer & value)
	{
	    char * src = buffer + readPos;
	    size32_t len = (size32_t)strlen(src);
	    CHECKREADPOS(len+1);
	    value.append(len, src);
	    readPos += (len+1);
	    return *this;
	}
	
	MemoryBuffer & MemoryBuffer::read(StringAttr & value)
	{
	    char * src = buffer + readPos;
	    size32_t len = (size32_t)strlen(src);
	    CHECKREADPOS(len+1);
	    value.set(src, len);
	    readPos += (len+1);
	    return *this;
	}
```

streeterd commented:
### Peformance Comparison
Comparing this change to 9.12.x I see:
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-GB style='font-family:Calibri;font-size:11.0pt'>
<!--StartFragment-->

<div style='direction:ltr'>


9.12.x | .14:35:40.962475   205472 UNK UNK Test Small Tree (wide): nodes 100; depth 1; content: text;   iterations: 1,000    Deserialized:   7,502 bytes       Runtime: 149 ms   14:35:41.174250   205472 UNK UNK Test Small Tree (deep): nodes 100; depth 1,000; content: text;   iterations: 1,000 Deserialized: 4,996 bytes       Runtime: 169 ms   .14:35:41.400241   205472 UNK UNK Test Medium Tree (wide): nodes 1,000; depth 1; content: text;   iterations: 100   Deserialized: 75,902   bytes      Runtime: 203 ms   14:35:41.574603   205472 UNK UNK Test Medium Tree (deep): nodes 1,000; depth 1,000; content:   text; iterations: 100          Deserialized: 51,744 bytes        Runtime: 135 ms   .14:35:47.431883   205472 UNK UNK Test Large Tree (wide): nodes 10,000; depth 1; content: text;   iterations: 100   Deserialized: 768,902   bytes     Runtime: 5587 ms   14:35:53.487046   205472 UNK UNK Test Large Tree (wide): nodes 10,000; depth 1; content:   binary; iterations: 100  Deserialized:   1,006,902 bytes   Runtime: 5714 ms   .14:36:04.101089   205472 UNK UNK Test Extra Large Tree (wide): nodes 100,000; depth 1; content:   text; iterations: 50     Deserialized:   7,788,902 bytes   Runtime: 8796 ms   14:36:16.829150   205472 UNK UNK Test Extra Large Tree (wide): nodes 100,000; depth 1; content:   binary; iterations: 50    Deserialized:   10,168,902 bytes  Runtime: 9951 ms   .14:36:35.015822   205472 UNK UNK Test Huge Tree (wide): nodes 1,000,000; depth 1; content:   text; iterations: 5   Deserialized:   78,888,902 bytes  Runtime: 13356 ms   14:36:56.259236   205472 UNK UNK Test Huge Tree (wide): nodes 1,000,000; depth 1; content:   binary; iterations: 5  Deserialized:   102,688,902 bytes Runtime: 14247 ms   .14:36:58.003954   205472 UNK UNK Test Deep Tree: nodes 5,000; depth 5,000; content: text;   iterations: 100        Deserialized:   267,744 bytes     Runtime: 707 ms   14:36:59.975904   205472 UNK UNK Test Wide Tree: nodes 5,000; depth 1; content: text;   iterations: 100     Deserialized:   383,902 bytes     Runtime: 1844 ms   14:37:00.867675   205472 UNK UNK Test Balanced Tree: nodes 5,000; depth 70; content: text;   iterations: 100        Deserialized:   250,622 bytes     Runtime: 796 ms   .14:37:10.571884   205472 UNK UNK Test Regression Dali Tree: nodes 23; depth 0; content: text;   iterations: 10     Deserialized:   33,152,343 bytes  Runtime: 6292 ms
-- | --
Step 1 | .12:06:20.656217   146337 UNK UNK Test Small Tree (wide): nodes 100; depth 1; content: text;   iterations: 1,000    Deserialized:   7,502 bytes       Runtime: 158 ms   12:06:20.826456   146337 UNK UNK Test Small Tree (deep): nodes 100; depth 1,000; content: text;   iterations: 1,000 Deserialized: 4,996 bytes       Runtime: 137 ms   .12:06:21.047421   146337 UNK UNK Test Medium Tree (wide): nodes 1,000; depth 1; content: text;   iterations: 100   Deserialized: 75,902   bytes      Runtime: 200 ms   12:06:21.223442   146337 UNK UNK Test Medium Tree (deep): nodes 1,000; depth 1,000; content:   text; iterations: 100          Deserialized: 51,744 bytes        Runtime: 139 ms   .12:06:27.128700   146337 UNK UNK Test Large Tree (wide): nodes 10,000; depth 1; content: text;   iterations: 100   Deserialized: 768,902   bytes     Runtime: 5653 ms   12:06:33.359175   146337 UNK UNK Test Large Tree (wide): nodes 10,000; depth 1; content:   binary; iterations: 100  Deserialized:   1,006,902 bytes   Runtime: 5896 ms   .12:06:44.744160   146337 UNK UNK Test Extra Large Tree (wide): nodes 100,000; depth 1; content:   text; iterations: 50     Deserialized:   7,788,902 bytes   Runtime: 9582 ms   12:06:57.663915   146337 UNK UNK Test Extra Large Tree (wide): nodes 100,000; depth 1; content:   binary; iterations: 50    Deserialized:   10,168,902 bytes  Runtime: 10479 ms   .12:07:16.285798   146337 UNK UNK Test Huge Tree (wide): nodes 1,000,000; depth 1; content:   text; iterations: 5   Deserialized:   78,888,902 bytes  Runtime: 13634 ms   12:07:37.708257   146337 UNK UNK Test Huge Tree (wide): nodes 1,000,000; depth 1; content:   binary; iterations: 5  Deserialized:   102,688,902 bytes Runtime: 14749 ms   .12:07:39.565898   146337 UNK UNK Test Deep Tree: nodes 5,000; depth 5,000; content: text;   iterations: 100        Deserialized:   267,744 bytes     Runtime: 772 ms   12:07:41.684767   146337 UNK UNK Test Wide Tree: nodes 5,000; depth 1; content: text;   iterations: 100     Deserialized:   383,902 bytes     Runtime: 1985 ms   12:07:42.716275   146337 UNK UNK Test Balanced Tree: nodes 5,000; depth 70; content: text;   iterations: 100        Deserialized:   250,622 bytes     Runtime: 931 ms   .12:07:53.296896   146337 UNK UNK Test Regression Dali Tree: nodes 23; depth 0; content: text;   iterations: 10    Deserialized:   33,152,343 bytes  Runtime: 7089 ms



</div>

<!--EndFragment-->
</body>

</html>

ghalliday commented on dali/base/dasds.cpp:
```diff
@@ -2468,7 +2468,8 @@ void CRemoteTreeBase::deserializeChildrenRT(MemoryBuffer &src)
 
 void CRemoteTreeBase::deserializeSelfRT(MemoryBuffer &mb)
```
This should have a context parameter as well, as should create(MemoryBuffer &) so that when it is called within deserialize (jptree.cpp(~3051) it can reuse the context.

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -3038,51 +3037,56 @@ void PTree::serialize(MemoryBuffer &tgt)
 
 void PTree::deserialize(MemoryBuffer &src)
```
deserialize also needs a context parameter - which would be created in createPTree()

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -3038,51 +3037,56 @@ void PTree::serialize(MemoryBuffer &tgt)
     unsigned pos = src.getPos();
     src.read(size);
-    if (value) delete value;
+    if (value)
```
trivial: no need to check for null before calling delete.

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -3038,51 +3037,56 @@ void PTree::serialize(MemoryBuffer &tgt)
-    if (_name[0]==0) 
-        setName(NULL);
+    if (deserializeContext.name.isEmpty())
+        setName(nullptr);
```
question: Is this needed I suspect it is already null.  Can it be replaced with dbgassertex(queryName()==nullptr);?

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -3112,3 +3112,31 @@ StringBuffer& sanitizeCommandArg(const char* arg, StringBuffer& sanitized)
+{
+    std::string s = std::to_string(value);
+    int n = s.length();
+    int firstGroupLen = n % GROUP_SIZE;
```
minor: I think ((n -1) % GROUP_SIZE) + 1 is simpler.

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -3112,3 +3112,31 @@ StringBuffer& sanitizeCommandArg(const char* arg, StringBuffer& sanitized)
+    for (i = firstGroupLen; i < n; i += GROUP_SIZE)
+    {
+        formatted.append(',');
+        for (int j = 0; j < GROUP_SIZE && (i + j) < n; j++) {
```
picky: The second test should not be needed.

ghalliday commented on testing/unittests/jlibtestsCustomPTree.xml:
```diff
@@ -0,0 +1 @@
+<root/>
```
I am not sure this should be checked in - normally unit tests would create any files that they then read.

ghalliday reviewed (changes requested):
Overall looks like a good step.  The PR should target master rather than 9.12.x since it is refactoring.
@jakesmith I had thought the idea was to implement a deserialize from an IBufferedRowStream first with the new changes, and then be able to call the new and old side by side for comparison.  Did that change?

streeterd committed:
1. Remove unit test based on checked in xml file
2. Change firstGroupLen computation in function formatWithCommas
3. Remove second test in third for loop in function formatWithCommas
4. Replace setName(nullptr) with dbgassertex
5. Remove null test before delete
6. Add context parameter to deserialize functions
7. Add context parameter to create(MemoryBuffer &) functions
8. Target PR to master

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

jakesmith reviewed (commented):
commenting purely to be off my gh review list for now.

ghalliday commented:
Closing.  I am not sure if any of this is still relevant.
